source/client.c

```diff
@@ -77,6 +77,38 @@ void aws_mqtt_client_clean_up(struct aws_mqtt_client *client) {
     AWS_ZERO_STRUCT(*client);
 }
 
+/* Reconnect helpers */
+static void s_attempt_reconnect(struct aws_task *task, void *userdata, enum aws_task_status status);
+
+int s_create_reconnect_task(struct aws_mqtt_client_connection *connection) {
+    AWS_ASSERT(!connection->reconnect_task);
+    connection->reconnect_task = aws_mem_calloc(connection->allocator, 1, sizeof(struct aws_mqtt_reconnect_task));
+    if (!connection->reconnect_task) {
+        AWS_LOGF_ERROR(AWS_LS_MQTT_CLIENT, "id=%p: Failed to allocate reconnect task", (void *)connection);
+        return AWS_OP_ERR;
+    }
+    aws_atomic_init_ptr(&connection->reconnect_task->connection_ptr, connection);
+    connection->reconnect_task->allocator = connection->allocator;
+    aws_task_init(&connection->reconnect_task->task, s_attempt_reconnect, connection->reconnect_task, "mqtt_reconnect");
+
+    return AWS_OP_SUCCESS;
+}
+
+static void s_destroy_reconnect_task(struct aws_mqtt_client_connection *connection) {
+    if (connection->reconnect_task) {
```

Comment:
Is this a safe use of atomics? shouldn't you be doing an atomic fetch for the null check?

Reply:
This is just a cut-and-paste of what existed before into a function so I can call it in two spots.